### PR TITLE
Lazy loading with flow compatibility

### PR DESCRIPTION
**Pull request by zertosh**
- Housekeeping in some of the root files.
- Bring back lazy loading - that works with flow!
- For completeness (since it doesn't cost anything) adde the `LInkedStateMixin`.
- Adds the missed `ReactDOMServer` - with flow compat! See https://github.com/facebook/flow/pull/1306 and `react-for-atom.js.flow`.

Tested against internal nuclide - no flow errors until I added:

```
[ignore]
.*/node_modules/react/.*
.*/node_modules/react-dom/.*
```

@jgebhardt, can add this project to travis please? Only owners can do it. Thanks!

cc: @ssorallen

### CODE REVIEW

File: react-for-atom.js.flow [hunk 1]
```diff
@@ -0,0 +1,43 @@
+'use strict';
```
Comment (ssorallen) [inline react-for-atom.js.flow:1]: This is a large file to have to maintain separately and correctly keep in sync. Is this really the best way?

**Comment by ssorallen** [comment]
> For completeness (since it doesn't cost anything) adde the LInkedStateMixin.

👿

File: .gitignore [hunk 2]
```diff
@@ -1,27 +1,2 @@
...
-
-# node-waf configuration
-.lock-wscript
-
-# Compiled binary addons (http://nodejs.org/api/addons.html)
-build/Release
-
-# Dependency directory
-# https://www.npmjs.org/doc/misc/npm-faq.html#should-i-check-my-node_modules-folder-into-git
+.DS_Store
```
Comment (ssorallen) [inline .gitignore:1]: This is an OS-specific file that should be ignored in your '~/.gitignore'. It's not generated by this project.

**Comment by ssorallen** [comment]
Lazy loading sounds good to me.

File: react-for-atom.js.flow (hunk 1 above)
Comment (zertosh) [inline react-for-atom.js.flow:1]: I couldn't get flow to work correctly with the getters. So we'd have to ditch that. I added tests to make sure that flow is behaving correctly.

File: .gitignore (hunk 2 above)
Comment (zertosh) [inline .gitignore:1]: That's why the commit message reads "gitignore for a mac-friendly node project" :stuck_out_tongue:. From a philosophical point of view, I agree with you. However, from a practical point of view, the reality is that macs are everywhere and this files are annoying.

If you feel strongly about it though, I'll remove it.

**Pull request merged**